### ISSUE TITLE
[RFC] Checkpoint design V2

### ISSUE THREAD

**Issue by andll** [MystenLabs/sui#5763]
# This document outlines new checkpoint design

## TLDR
The new design significantly simplifies checkpoint protocol by putting all certificates in the consensus. The checkpoint content is then implicitly derived from consensus sequence, without need to form fragments and spanning trees.

## TODO

[ ] Submit all certificates to consensus
[ ] Reduce number of consensus transactions for each certificate
[ ] Derive checkpoint content, submit signatures to consensus
[ ] Aggregate signatures in checkpoint certificates
[ ] Shutdown accepting new certificates, send EndOfPublish message
[ ] Process 2f+1 EndOfPublish, make new epoch
[ ] Edge cases with detecting new epoch, switching narwhal, etc
[ ] Persistent `ConsensusWriter` -or- strong narwhal worker guarantees 

## Motivation

- Simplifying checkpoint mechanism compared to fragments approach.
- Unifies data dissemination for validators - narwhal is used for data dissemination and we can  retire gossip protocol between validators.
- The checkpoint latency is reduced dramatically (lower bound is narwhal commit latency + RTT), which allows to use checkpoints for full node sync.
- Catch up to checkpoint within epoch becomes easier for validator. Basically you just follow narwhal stream and there is no separate mechanism for that.
- Narwhal stream now have a lot of information - certificates, checkpoint signatures, etc. A lot of it can be used for calculating validator rewards.

## Guarantees

- Checkpoints is a list of effect digests in casual order, signed by f+1 validators
- If `handle_certificate` on validator returned `SignedEffects`, the validator will attempt(*) to include transaction in checkpoint before end of epoch
    - Consequence: when gathered 2f+1 SignedEffects, transaction is final, e.g. guaranteed to appear in the checkpoint before end of epoch
- For owned object transactions `handle_certificate` does not block on consensus ⇒ “low owned object latency” requirement

(*) The epoch might end before all validators submit all the transactions they executed. However, we require that at least 2f+1 validators submitted all the executed transactions before end of epoch.

## Algorithm

1. **Submission**. `handle_certificate` includes all certificates to the consensus(rn only doing it for shared objects)
    - **Optimization 1**: we [[implement mechanism](https://www.notion.so/Checkpointing-via-Sequencing-Design-Notes-e658641c5694475ab64c338984cedeab)](https://www.notion.so/Checkpointing-via-Sequencing-Design-Notes-e658641c5694475ab64c338984cedeab) to minimize number of repeating certificates in consensus - ideally only 1 cert per user transaction, but might want more for censorship/MEV resistance.
    - **Optimization 2**: optimizing narwhal worker exchange to reduce traffic and not copy over entire certificates - attempt to exchange digests first, then exchange missing certificates only
    - Note that both optimizations would be needed for shared transactions, but putting owned object certs in consensus makes them more desired early on

**Discussion**: why use certificates and not effect digest

- Effect digests are shorter and help reduce narwal bandwidth. However it is not clear when narwal bandwidth will be a bottleneck, and when it is, **there is solution for this problem** (Optimization 2)
- Putting certificates allows to **reuse some of the trickier parts of the implementation**, for ex. deciding when/who puts certificate in consensus to reduce duplicates. In other words, we just do same thing for owned objects that we did for shared, instead of adding entirely new consensus interaction flow
- Most importantly, using certificates for checkpoints allows to **reduce checkpoint latency**
    - For owned object we put certificate parallel to execution, if we were putting digest we would have to wait for execution first.
    - For shared objects latency impact is even more pronounced - if we go effect digest route, we would have to **pay consensus latency twice** before transaction is included in the checkpoint - first we need to sequence certificate and only after that we can put effect digest in consensus for checkpoint
1. **Checkpoint selection rule**. Checkpoint is built based on a set of certificates in the certain range of narwhal stream. Validators have implicit agreement on what position in the narwhal stream checkpoint is generated at.
    - For example validators can implicitly agree to create checkpoint on the boundary of every 10th narwhal certificate.
2. **Checkpoint formation rule.** Rule (2) gives all validators consistent list of certificates that are basis for forming a checkpoint. However, the list of certificates is neither complete(it might be missing casual dependencies) nor it is casually ordered. Additional rule is required to form a consistent list of transactions for the checkpoints based on the list of certificates. We can chose this rule as following: *we add casual dependencies to the list, remove transactions previously included in checkpoint, order transactions in casual order* using the rule that yields same results on all validator (it is possible to produce such a rule).
    - Alternative rule can be excluding certificates that do not have casual dependencies in consensus yet, until those dependencies are submitted. I think the mentioned above “add missing” rule might be easier to implement
3. **Signing checkpoint**. When validator receives narwhal transaction in a position that requires forming a checkpoint, it waits for all the transactions in that checkpoint to be executed, forms checkpoint locally and submit `(CheckpointHeight, CheckpointSignature)` to the consensus. This is done in the process asynchronous to critical consensus task that sets shared locks. Note that *checkpoint content is only shared implicitly*. Some additional information (like list of digests) could be included in consensus to simplify debugging initially, but is not required for checkpoint process to work.
    - Why put signatures in narwal?
        - Pin signatures for checkpoint certificate
        - Can be used for reward calculation
        - No need to addition mechanism
4. **Certifying**. The certificate for the checkpoint can be read by following consensus stream and combining `CheckpointSignature` on the same height. Consistency of rule (2) guarantees that validators implicitly agree on what is the content of the checkpoint on the given height.
5. **Reconfiguration**. Special care is needed to support guarantee (2) at epoch boundary. 
    
    **Context**: In the current implementation we denote certain **checkpoint height C** at which epoch ends, validators stop accepting new certificates after reaching C-1. We can still support this rule, by augmenting selection rule (2) for the last checkpoint. The main issue with having a fixed checkpoint height C though is that it means that **last checkpoint has an unbounded number of digests in it**. Note that this is fundamental issue(it is present in the current scheme as well) - since there is no bound of number of transactions not yet in the checkpoint and there is fixed checkpoint height, the number of digests in the last checkpoint is also unbounded. **Having unbounded data structures in the system is a liveness problem**, so it must be avoided. With current system it would result in arbitrary large fragments at the end of epoch. Below is alternative proposal that does not require ending epoch at fixed height and therefore allows bounded number of transactions in the checkpoint.
    
    - Validator have some **loosely defined condition for when the epoch ends** - for example certain time of day according to their local clock or reaching certain non-terminal checkpoint height, etc.
    - When **condition is reached locally**, validator stops accepting new certificates from user but maintain a background process for pushing pending processed certificates to the consensus and participates in checkpoint signing as usual
    - When all pending certificates are pushed to the consensus, validator submits special “**end of publish**” transaction to the narwhal. After this certificates published by this validator are ignored (correct validator does not submit anything at this point, but quirks in narwhal can still lead to some certificates appear to be published by the validator after the “end of publish” message). The pending submission process on the given validator terminates here, but validator still follows narwhal stream and participate in signing checkpoints.
        - This special “end of publish” transaction could simply be each validator’s signed epoch change transaction. Right now the epoch change transaction includes 1) incrementing the on-chain epoch number, 2) processing pending committee member and stake changes, as well as 3) distributing staking rewards coming from gas charges of all transactions in the old epoch. We change the epoch change transaction under checkpoint v2 to work as follows:
            - Each validator forms and signs an epoch change transaction that only does 1) and 2), and submits the signed transaction to Narwhal.
            - The rewards distribution part of the process can happen asynchronously as one of the first transactions in the new epoch. We don’t have to execute this transaction as the very first transaction of the new epoch because the compounded staking rewards are not counted as stakes until after the new epoch has passed. Therefore, when we execute the rewards distribution transaction doesn’t affect validators’ staking powers as longs it happens before the new epoch ends.
    - When **threshold number**(2f+1) of “end of publish” messages appear in the consensus epoch is considered “**closed**” - any certificates in the consensus after this are ignored. We can denote position at narwhal stream where threshold number of “end of publish” message is seen as “**closed epoch boundary**”.
    - After epoch is closed all certificates in the narwhal are ignored(and correct validator stop submitting new transactions). Validators still need to submit signature(s) for checkpoints(s) up until position of “closed epoch” in narwhal sequence. When enough validators submit signatures to form **certified checkpoint at “closed epoch boundary”** the last checkpoint is certified and epoch has ended.

## Open questions

**How does this affect validator gossip?** Do we need additional mechanism to gossip certificates when all of them are in consensus?

- Delay of narwal is higher then delay of optimistic gossip - need to figure out if this is important in practice
    - If the goal is to handle certificates ASAP we actually don’t need to wait for ~~narwhal~~ bullshark commit and take certificates when they arrive at the narwhal worker. However, such optimistically arrived certificates do need to go through generic handle_certificate checks that rejects certs after issuing “end of publish” message.

**Do we need a separate story for “catch up to a checkpoint”** within the epoch? Is simply following narwhal enough(at least to start with)?

- I think the answer is no
